### PR TITLE
[examples] Added error handling to the trib app for models and templates for errors and when no results are returned.

### PR DESCRIPTION
**Pull request by zhouyaoji**
The models of the mojits Blog, Calendar, Gallery, Github, and Youtube would occasionally not return results because of an issue with YQL or the original data source. Because the models didn't handle this case, the app would fail. Two shared templates were added that mojits could use to render HTML when there was an error or no results.
In addition, an error in the Calendar model that incorrectly parsed a string to extract a date was corrected.

### CODE REVIEW

**Comment by zhouyaoji** [comment]
This PR is intended to fix issue #1253.

File: examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js [hunk 1]
```diff
@@ -38,12 +39,13 @@ YUI.add('Blog', function (Y, NAME) {
...
                 ac.assets.addCss('./index.css');
-
+                Y.log(data.view, "info");
+                Y.log(data.error, "info");
                 // populate blog template
                 ac.done({
                     title: title,
                     results: data
-                });
+                }, data.view);
```
Comment (imalberto) [inline examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js]: Models should not be telling controllers which view to select. Instead, models should return either results, or error, and controller should take appropriate action.

File: examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js [hunk 2]
```diff
@@ -33,9 +34,14 @@ YUI.add('BlogModelYQL', function (Y, NAME) {
         onDataReturn: function (cb, result) {
             Y.log("blog.server onDataReturn called");
             var results = [];
-            if (result.error === undefined) {
+            results.view = "index";
+            if (result.query && result.query.results === null) {
```
Comment (imalberto) [inline examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js]: Only check for truthy or falsy

File: examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js [hunk 3]
```diff
@@ -38,15 +39,21 @@ YUI.add('Blog', function (Y, NAME) {
                 //Y.log(data);
                 // add mojit specific css
                 ac.assets.addCss('./index.css');
-
-                // populate blog template
-                ac.done({
-                    title: title,
-                    results: data
-                });
+                Y.log("In Blog controller: ", "info");
```
Comment (caridy) [inline examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js]: "info", NAME);

File: examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js [hunk 4]
```diff
@@ -38,15 +39,21 @@ YUI.add('Blog', function (Y, NAME) {
...
-                ac.done({
-                    title: title,
-                    results: data
-                });
+                Y.log("In Blog controller: ", "info"); 
+                if(data.error) {
+                   // Error was found, so render `error` template.
+                   ac.done({ 
+                      title: title, 
+                      results: data
```
Comment (caridy) [inline examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js]: do the `error` template requires `results`? or just the title?

Comment (caridy) [inline examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js]: I will argue that the right thing here is to do ac.error(); and let the parent to deal with the error, but I might not have the full info here.

File: examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js [hunk 5]
```diff
@@ -37,19 +38,17 @@ YUI.add('BlogModelYQL', function (Y, NAME) {
 
                 if (result && result.query && result.query.results && result.query.results.item) {
                     results = result.query.results.item;
+                    Y.blogData = results;
```
Comment (caridy) [inline examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js:42]: we should NOT encourage people to do this kind of cache. If this cache is absolutely necesary, I will recommend do `YUI.Env.*`, which explicitly state that it is global.

File: examples/developer-guide/dashboard/trib/mojits/Blog/views/index.hb.html [hunk 6]
```diff
@@ -2,14 +2,17 @@
     <div class="mod" id="blog">
         {{> widget_heading}}
         <div class="inner">
-            <ul>
-                {{#results}}
+            {{#each results}}
+                <ul>
```
Comment (caridy) [inline examples/developer-guide/dashboard/trib/mojits/Blog/views/index.hb.html]: Each result will be wrap by an UL? sounds weird.

File: examples/developer-guide/dashboard/trib/mojits/Calendar/controller.server.js [hunk 7]
```diff
@@ -30,17 +31,22 @@ YUI.add('Calendar', function (Y, NAME) {
...
-
-                // populate blog template
-                ac.done({
-                    title: "YUI Calendar Info",
-                    results: data
-                });
+                if(data.error) {
+                    // Error was found: render `error` template.
+                    ac.done({ 
+                        title: title,
```
Comment (caridy) [inline examples/developer-guide/dashboard/trib/mojits/Calendar/controller.server.js]: same as above!

File: examples/developer-guide/dashboard/trib/mojits/Calendar/controller.server.js [hunk 8]
```diff
@@ -23,6 +23,7 @@ YUI.add('Calendar', function (Y, NAME) {
          */
         index: function (ac) {
             ac.models.get('CalendarModelYQL').getData({}, function (data) {
+                var title = "YUI Calendar Info";
```
Comment (caridy) [inline examples/developer-guide/dashboard/trib/mojits/Calendar/controller.server.js:26]: Isn't this suppose to come from the lang bundle?

File: examples/developer-guide/dashboard/trib/mojits/Calendar/models/calendar.server.js [hunk 9]
```diff
@@ -27,6 +28,7 @@ YUI.add('CalendarModelYQL', function (Y, NAME) {
         },
         onDataReturn: function (cb, result) {
             Y.log("calendar.server onDataReturn called");
```
Comment (caridy) [inline examples/developer-guide/dashboard/trib/mojits/Calendar/models/calendar.server.js]: proper log signature across all these files please.

File: examples/developer-guide/dashboard/trib/mojits/Calendar/models/calendar.server.js [hunk 10]
```diff
@@ -36,32 +38,25 @@ YUI.add('CalendarModelYQL', function (Y, NAME) {
...
+                    //Y.log(results[0].entry.summary.content);
+                    Y.Array.each(results, function (val, key, obj) {
+                        Y.log(val.entry.summary.content);
+                        var tempDate = val.entry.summary.content;
+                        // Strip off 'br', 'When:'' and 'to' elements to get date
+                        tempDate = tempDate.split("<")[0].split("When:")[1]; 
+                        val.entry.summary.content = tempDate;
+                        Y.log(val.entry.summary.content);
+                    });
+                    Y.calendarData = results;
```
Comment (caridy) [inline examples/developer-guide/dashboard/trib/mojits/Calendar/models/calendar.server.js:51]: same as above.

File: examples/developer-guide/dashboard/trib/mojits/Calendar/views/index.hb.html [hunk 11]
```diff
@@ -2,11 +2,13 @@
     <div class="mod" id="calendar">
         {{> widget_heading}}
         <div class="inner">
-            <ul>
-                {{#results}}
+            {{#each results}}
+                <ul>
```
Comment (caridy) [inline examples/developer-guide/dashboard/trib/mojits/Calendar/views/index.hb.html]: same as above

**Comment by caridy** [comment]
Ok, I added a lot of comments, but I guess I got confused since that was not the original purpose of the PR, although some of them are relevant for the error manipulation. Feel free to ignore the others.

File: examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js (hunk 4 above)
Comment (zhouyaoji) [inline examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js]: The returned `data` object will have the results or an `error` object. The `error` template displays the error, which is not desired effect in production, but for an example app, it makes it easier to view errors. If you use `ac.error`, the child mojit's template will simply not be rendered, but the error will show up in the server logs. Just using `ac.error` might be the better solution though.

File: examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js (hunk 5 above)
Comment (zhouyaoji) [inline examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js:42]: The caching implementation was created by the YUI team. I don't include it in the tutorial. Can you file a separate issue for this? Thanks.

File: examples/developer-guide/dashboard/trib/mojits/Calendar/controller.server.js (hunk 8 above)
Comment (zhouyaoji) [inline examples/developer-guide/dashboard/trib/mojits/Calendar/controller.server.js:26]: The `lang` bundle is just used by the `PageLayout` mojit for the top-level heading as an example. The child mojits do not perform localization.

File: examples/developer-guide/dashboard/trib/mojits/Blog/views/index.hb.html (hunk 6 above)
Comment (zhouyaoji) [inline examples/developer-guide/dashboard/trib/mojits/Blog/views/index.hb.html]: The `index` template also handle the case where no results are shown; thus, I didn't want to have a `<ul>` if there were no results, but it shouldn't wrap each result with a `<ul>`. but I'll confirm and make changes if needed.

File: examples/developer-guide/dashboard/trib/mojits/Calendar/models/calendar.server.js (hunk 9 above)
Comment (zhouyaoji) [inline examples/developer-guide/dashboard/trib/mojits/Calendar/models/calendar.server.js]: Will do. Thanks!

**Comment by zhouyaoji** [comment]
@Caridy Thanks for your comments! I'll make the log statements more consistent, confirm that I'm not wrapping each result in a `<ul>` element, and have the children use an `ac.error` instead of using an `error` template. I'm going to leave the caching mechanism in place. We should file an issue to correct that.

File: examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js (hunk 5 above)
Comment (imalberto) [inline examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js:42]: @zhouyaoji You could cache the results by using a variable local to the module vs attaching it to the `Y` instance.

**Comment by zhouyaoji** [comment]
@caridy I removed the `error` template and used `ac.error` instead, which will just show the errors in the server logs. I also cleaned up the `Y.log` statements, and corrected the templates so that not each result is wrapped in `<ul>` tags--you were correct as usual. Anyway, let me know if I can go ahead and merge the changes. Again, we can file an issue to fix the caching next. Thanks again!

File: examples/developer-guide/dashboard/trib/mojits/Gallery/models/gallery.server.js [hunk 12]
```diff
@@ -18,37 +19,32 @@ YUI.add('GalleryModelYQL', function (Y, NAME) {
                     },
                     cookedQuery = Y.Lang.sub(query, queryParams);
 
-                    //Y.log("cookedQuery: " + cookedQuery);
-
+                    //Y.log("cookedQuery: " + cookedQuery, "info", NAME);
```
Comment (imalberto) [inline examples/developer-guide/dashboard/trib/mojits/Gallery/models/gallery.server.js:22]: `if (!result.error) {` would work too

File: examples/developer-guide/dashboard/trib/mojits/Gallery/models/gallery.server.js [hunk 13]
```diff
@@ -18,37 +19,32 @@ YUI.add('GalleryModelYQL', function (Y, NAME) {
...
+                Y.log(result, "info", NAME);
 
                 if (result && result.query && result.query.results && result.query.results.json) {
                     results = result.query.results.json;
                     results.json = results.json.slice(0, itemLimit);
+                    Y.galleryData = results;
+                    Y.galleryCacheTime = new Date().getTime();
+                } else {
+                    results = new Error({ error: { description: "The returned response was malformed." }}); 
                 }
```
Comment (imalberto) [inline examples/developer-guide/dashboard/trib/mojits/Gallery/models/gallery.server.js:42]: Probably want to follow the node convention:
- if error: `cb(results.error)`
- if no error: `cb(null, results)`

File: examples/developer-guide/dashboard/trib/mojits/Youtube/models/youtube.server.js [hunk 14]
```diff
@@ -20,38 +20,30 @@ YUI.add('YoutubeModelYQL', function (Y, NAME) {
...
-                //Y.log("results: ");
-                //Y.log(results);
-
-
-                cb(results);
             } else {
-                cb(result.error);
+                results = result;
             }
+            cb(results);
```
Comment (imalberto) [inline examples/developer-guide/dashboard/trib/mojits/Youtube/models/youtube.server.js]: Same as above. Follow node callback convention.

File: examples/developer-guide/dashboard/trib/mojits/Youtube/controller.server.js [hunk 15]
```diff
@@ -43,20 +41,23 @@ YUI.add('Youtube', function (Y, NAME) {
...
-                    title: "YUI YouTube Videos",
-                    results: res
-                });
+                Y.log("youtubmojit results:", "info", NAME);
+                Y.log(res, "info", NAME);
+                if (data.error) {
+                    ac.error(data);
+                } else {
+                    // Create data structure from Web service response.
+                    res = youtubeMap(ac, data);
```
Comment (imalberto) [inline examples/developer-guide/dashboard/trib/mojits/Youtube/controller.server.js]: It's better not to pass `ac` to this function. This function is only iterating through the data. Makes the intent clearer.

File: examples/developer-guide/dashboard/trib/mojits/Github/models/yql.server.js [hunk 16]
```diff
@@ -24,33 +23,31 @@ YUI.add('StatsModelYQL', function (Y, NAME) {
...
+                } 
+                Y.log("github: results.json:", "info", NAME);
+                Y.log(results, "info", NAME);
 
-                cb(results);
             } else {
-                cb(result.error);
+                results = result;
             }
+            cb(results);
```
Comment (imalberto) [inline examples/developer-guide/dashboard/trib/mojits/Github/models/yql.server.js]: `cb(err)` vs `cb(null, results)`.

File: examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js [hunk 17]
```diff
@@ -16,45 +17,42 @@ YUI.add('BlogModelYQL', function (Y, NAME) {
...
-
-
-                cb(results);
+                Y.log("result.query.results.item = results: ", "info", NAME);
+                Y.log(results, "info", NAME);
             } else {
-                cb(result.error);
+                results = result;
             }
+            cb(results);
```
Comment (imalberto) [inline examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js]: Same here with passing the error as 1st arg, but if no error, the results should be the 2nd arg.

**Comment by imalberto** [comment]
Done with review.

File: examples/developer-guide/dashboard/trib/mojits/Gallery/models/gallery.server.js (hunk 12 above)
Comment (zhouyaoji) [inline examples/developer-guide/dashboard/trib/mojits/Gallery/models/gallery.server.js:22]: Thanks, I went with your suggestion.

**Comment by imalberto** [comment]
@zhouyaoji LGTM. Please check the build failures on Travis that it's not related to the PR. Otherwise +1

**Pull request merged**